tools/ctl/ctl.go

```diff
@@ -294,88 +472,38 @@ var runBenchmark = &cobra.Command{
+				WorkerConfig: wrkConf,
 			}
-			close(jobs)
-		}()
-		limiterMap := map[string]<-chan time.Time{
-			"pypi":  time.Tick(time.Second),
-			"npm":   time.Tick(2 * time.Second),
-			"maven": time.Tick(2 * time.Second),
-			// NOTE: cratesio needs to be especially slow given our registry API
-			// constraint of 1QPS. At minimum, we expect to make 4 calls per test.
-			"cratesio": time.Tick(8 * time.Second),
-		}
-		var totalErrors int
-		var aggErrors []string
-		var wg sync.WaitGroup
-		for i := 0; i < *maxConcurrency; i++ {
-			wg.Add(1)
-			go func() {
-				defer wg.Done()
-				if mode == firestore.SmoketestMode {
-					// First, warm up the instances to ensure it can handle actual load.
-					// Warm up requires the service fulfill sequentially successful version
-					// requests (which hit both the API and the builder jobs).
-					for i := 0; i < 5; {
-						_, err := getExecutorVersion(ctx, idclient, apiURL, "build-local")
-						if err != nil {
-							i = 0
-						} else {
-							i++
-						}
-					}
-				}
-				// Second, start triggering rebuilds.
-				for j := range jobs {
-					var reqs []*http.Request
-					if mode == firestore.SmoketestMode {
-						reqs = append(reqs, makeHTTPRequest(ctx, apiURL.JoinPath("smoketest"), schema.SmoketestRequest{
-							Ecosystem: rebuild.Ecosystem(j.Ecosystem),
-							Package:   j.Name,
-							Versions:  j.Versions,
-							ID:        run,
-						}))
-					} else if mode == firestore.AttestMode {
-						for _, v := range j.Versions {
-							reqs = append(reqs, makeHTTPRequest(ctx, apiURL.JoinPath("rebuild"), schema.RebuildPackageRequest{
-								Ecosystem: rebuild.Ecosystem(j.Ecosystem),
-								Package:   j.Name,
-								Version:   v,
-								ID:        run,
-							}))
-						}
-					}
-					for _, req := range reqs {
-						// Wait for a tick from the limiter.
-						<-limiterMap[j.Ecosystem]
-						resp, err := idclient.Do(req)
-						if err != nil {
-							totalErrors++
-							aggErrors = append(aggErrors, errors.Wrap(err, "sending request").Error())
-							continue
-						}
-						if resp.StatusCode != 200 {
-							totalErrors++
-							aggErrors = append(aggErrors, errors.Wrap(errors.New(resp.Status), "request").Error())
-						}
-					}
-					bar.Increment()
-				}
-			}()
 		}
-		wg.Wait()
+		verdictChan := make(chan schema.Verdict)
+		go ex.Process(ctx, verdictChan, set.Packages)
+		var verdicts []schema.Verdict
+		for v := range verdictChan {
+			verdicts = append(verdicts, v)
+		}
 		bar.Finish()
-		log.Printf("Completed rebuilds for %d artifacts...\n", set.Count)
-		log.Printf("Total errors: %d\n", totalErrors)
-		for _, e := range aggErrors {
-			log.Println(e)
+		sort.Slice(verdicts, func(i, j int) bool {
+			return fmt.Sprint(verdicts[i].Target) > fmt.Sprint(verdicts[j].Target)
+		})
+		log.Printf("Triggering rebuilds on executor version '%s' with ID=%s...\n", executor, run)
+		for _, v := range verdicts {
+			fmt.Printf("%s,%s", fmt.Sprint(v.Target), v.Message)
```

Comment:
maybe leave a TODO to support more optional (?) data? we may want other parts of the verdict here e.g. strategy.